book/src/datapipes.py

```diff
@@ -0,0 +1,203 @@
+        
+    def __init__(self, ds, selected_regions=None, preload=False):
+        super().__init__()
+        self.ds = ds
+        self.selected_regions = self.to_sequence(selected_regions)
+        self.preload = preload
+        
+    def to_sequence(self, seq):
+        if isinstance(seq, str):
+            return (seq, )
+        return seq
+
+    def __iter__(self):
+        if not self.selected_regions:
+            yield self.ds
+        else:
+            for region in self.selected_regions:
+                self.selected_ds = select_region(self.ds, region)
+                if self.preload:
+                    self.selected_ds = self.selected_ds.load()
+                yield self.selected_ds
+            
+            
+def filter_batch(batch):
+    return batch.where(batch['mask']>0, drop=True)
+
+
+def transform_batch(
+    batch, 
+    scale_means=scale_means, 
+    scale_stds=scale_stds
+):
+    batch = (batch - scale_means) / scale_stds
+    return batch
+
+
+def untransform_batch(
+    batch, 
+    scale_means=scale_means, 
+    scale_stds=scale_stds
+):
+    batch = (batch * scale_stds) + scale_means
+    return batch
+
+
+def stack_split_convert(
+    batch, 
+    in_vars, 
+    out_vars, 
+    in_selectors={},
+    out_selectors={},
+    min_samples=50,
+    dtype=torch.float32
+):
+    if len(batch['sample']) > min_samples:
+        x = (batch[in_vars]
+                 .to_array()
+                 .transpose('sample', 'time', 'variable')
+                 .isel(**in_selectors))
+        y = (batch[out_vars]
+                  .to_array()
+                  .transpose('sample', 'time', 'variable')
+                  .isel(**out_selectors))
+        x = torch.tensor(x.values).to(dtype)
+        y = torch.tensor(y.values).to(dtype)
+    else:
+        x, y = torch.tensor([], dtype=dtype), torch.tensor([], dtype=dtype)
+    return x, y
+
+
+def make_data_pipeline(
+    ds, 
+    regions, 
+    input_vars, 
+    output_vars,
+    input_sequence_length,
+    output_sequence_length,
+    batch_dims,
+    input_overlap,
+    preload=False,
+    min_samples=50,
+    filter_mask=True,
+    dtype=torch.float32,
+    **kwargs
+):
+    # Preamble: just set some stuff up
+    output_selector = {'time': slice(-output_sequence_length, None)}
+    input_dims={'time': input_sequence_length}
+    varlist = ['mask'] + input_vars + output_vars
+    convert = partial(
+        stack_split_convert, 
+        in_vars=input_vars, 
+        out_vars=output_vars, 
+        out_selectors=output_selector,
+        min_samples=min_samples,
+        dtype=dtype,
+    )
+    # Chain together the datapipe
+    dp = RegionalSubsetterPipe(
+        ds[varlist], selected_regions=regions, preload=preload
+    )
+    dp = dp.slice_with_xbatcher(
+        input_dims=input_dims,
+        batch_dims=batch_dims,
+        input_overlap=input_overlap,
+        preload_batch=False
+    )
+    if filter_mask:
+        dp = dp.map(filter_batch)
```

Comment:
Quick comment that it might be semantically better to use https://pytorch.org/data/0.6/generated/torchdata.datapipes.iter.Filter.html for filtering out patches with NaNs. I.e.

```
dp = dp.filter(filter_fn=filter_batch)
```

That said, I haven't check your data pipeline too closely, but my impression with using `.map(filter_batch)` was that you were zero-ing out the data rather than removing them?